core/src/main/scala/kafka/log/FileMessageSet.scala

```diff
@@ -126,26 +126,136 @@ class FileMessageSet private[kafka](@volatile var file: File,
    * @param targetOffset The offset to search for.
    * @param startingPosition The starting position in the file to begin searching from.
    */
-  def searchFor(targetOffset: Long, startingPosition: Int): OffsetPosition = {
+  def searchForOffset(targetOffset: Long, startingPosition: Int): OffsetPosition = {
+    var foundOffset = -1L
+    var foundPosition = -1
+    def handleMessage(offset: Long, position: Int, magic: Byte, timestamp: Long): Boolean = {
+      if (offset >= targetOffset) {
+        foundOffset = offset
+        foundPosition = position
+        true
+      } else false
+    }
+
+    try scanAndHandleMessages(startingPosition, handleMessage)
+    catch {
+      case e: Exception => throw new KafkaException(s"Failed to search for offset $targetOffset due to", e)
+    }
+
+    if (foundOffset < 0) null
+    else OffsetPosition(foundOffset, foundPosition)
+  }
+
+  /**
+   * Search forward for the message whose timestamp is greater than or equals to the target timestamp.
+   * The method will return
+   * <ul>
+   * <li>None if there is no message after the starting position.
+   * <li>TimestampOffset( largest_timestamp_checked_before_target_timestamp, the_next_offset_to_read ) otherwise.
+   * This is enssentially a time index entry built for all the messages from starting position until it finds
+   * the target timestamp or reaches the end of the segment. If the message whose timestamp is greater than
+   * or equals to the target timestamp is found, the next_offset_to_read will be the offset of that message,
+   * otherwise, it will be the next offset of the last offset in this segment.
```

Comment:
The description starting from line 155 seems redundant. The description in line 150 and 151 explains this method pretty well.